### PR TITLE
reduce simdjson memory consuming

### LINKED ISSUE #2344: memory consuming of json stream load
### Background
https://github.com/StarRocks/starrocks/blob/389324198bce9cd4079bcb1e779be5360971b870/be/src/exec/vectorized/json_scanner.cpp#L704
To support `ndjson` iteration, we use `simdjson::ondemand::parser::iterate_many`.
A 200MB buffer is pre-allocated to support big document (such as array) parsing, which brings huge memory consuming in case with high concurrency.  

### Solution
The pre-allocation could be eliminated by using `simdjson::ondemand::parser::iterate` to iterate json in array, and using `simdjson::ondemand::parser::iterate_many` to iterate json in `ndjson`. We could determine the input type according to the "strip_outer_array" header.

### PR DESCRIPTION
**Pull request by rickif**
This PR is to reduce the memory consumption in json stream load, includes following modifications:
1. using `simdjson::ondemand::parser::iterate` to parse json array
2. using `simdjson::ondemand::parser::iterate_many` to parse ndjson
3. eliminating the pre-allocated buffer for `iterate_many`

With this optimization, in my test case of routine load with 400 concurrencies, the  `Virtual Memory Size` would be reduced from 150GB to 8GB

### CODE REVIEW

**Comment by rickif** [comment]
## Memory Consuming (before optimizing)
image
## Memory Consuming (after optimizing)
image

File: be/src/exec/vectorized/json_scanner.h [hunk 1]
```diff
@@ -134,4 +117,43 @@ class JsonReader {
...
+public:
+    Status parse(uint8_t* data, size_t len, size_t allocated) override;
+    Status get_next(simdjson::ondemand::object* row) override;
+
+private:
+    uint8_t* _data;
+    simdjson::ondemand::parser _parser;
+
+    // _first_shot denotes wheather the first element of the document stream would be returned.
+    bool _first_shot = false;
```
Comment (chaoyli) [inline be/src/exec/vectorized/json_scanner.h:138]: When the first element do not return?

File: be/src/exec/vectorized/json_scanner.cpp [hunk 2]
```diff
@@ -683,37 +516,43 @@ Status JsonReader::_filter_row_with_jsonroot(simdjson::ondemand::object* row) {
...
+
 #endif
 
-    if (err) {
-        std::string err_msg = strings::Substitute("Failed to parse string to json. code=$0, error=$1", err,
-                                                  simdjson::error_message(err));
-        return Status::DataQualityError(err_msg.c_str());
+    // If _strip_outer_array == true, we try to iterate the json as array.
+    if (_scanner->_strip_outer_array) {
+        _parser.reset(new JsonArrayParser);
```
Comment (chaoyli) [inline be/src/exec/vectorized/json_scanner.cpp:549]: What's the difference between JsonArrayParser and JsonDocumentStreamParser?

File: be/src/exec/vectorized/json_scanner.h (hunk 1 above)
Comment (rickif) [reply be/src/exec/vectorized/json_scanner.h:138]: The first element is returned when the `get_next()` is called for the first time after the `parse()`.

Comment (rickif) [reply be/src/exec/vectorized/json_scanner.h:139]: `_first_shot ` is the flag to tell us weather to forward the iterator.

File: be/src/exec/vectorized/json_scanner.cpp (hunk 2 above)
Comment (rickif) [reply be/src/exec/vectorized/json_scanner.cpp:549]: `JsonArrayParser` using `simdjson::ondemand::parser::iterate` to parse json array, which only supports parsing single document in streaming.
`JsonDocumentStreamParser` using `simdjson::ondemand::parser::iterate_many` to parse ndjson.

Comment (chaoyli) [reply be/src/exec/vectorized/json_scanner.cpp:549]: OK. Do you use strip_outer_array to check whether the input is a ndjson or not?

File: be/src/exec/vectorized/json_scanner.h (hunk 1 above)
Comment (chaoyli) [reply be/src/exec/vectorized/json_scanner.h:139]: Does the _first_shot is necessary? Could we use the get_next() to forward the iterator automatically?

File: be/src/exec/vectorized/json_scanner.cpp (hunk 2 above)
Comment (rickif) [reply be/src/exec/vectorized/json_scanner.cpp:549]: The `strip_outer_array` is checked before the `JsonParser` is constructed.Then, the json input format is checked in `JsonParser::parser()`.

File: be/src/exec/vectorized/json_scanner.h (hunk 1 above)
Comment (rickif) [reply be/src/exec/vectorized/json_scanner.h]: The `get_next()` is divided into 2 parts: `next()` and `get()`, and the `_first_shot ` is removed.

File: be/src/exec/vectorized/json_scanner.cpp [hunk 3]
```diff
@@ -656,4 +669,17 @@ Status JsonArrayParser::get_next(simdjson::ondemand::object* row) {
     }
 }
 
+Status JsonArrayParser::next() {
```
Comment (chaoyli) [inline be/src/exec/vectorized/json_scanner.cpp:673]: advance() is a better name.

File: be/src/exec/vectorized/json_scanner.cpp [hunk 4]
```diff
@@ -627,21 +648,13 @@ Status JsonArrayParser::parse(uint8_t* data, size_t len, size_t allocated) {
     return Status::OK();
 }
 
-Status JsonArrayParser::get_next(simdjson::ondemand::object* row) {
+Status JsonArrayParser::get(simdjson::ondemand::object* row) {
```
Comment (chaoyli) [inline be/src/exec/vectorized/json_scanner.cpp:652]: get_current() may be a better name.

File: be/src/exec/vectorized/json_scanner.cpp [hunk 5]
```diff
@@ -208,40 +208,29 @@ Status JsonScanner::_construct_cast_exprs() {
...
+            const char* cstr = path.get_c_str();
 
-        std::vector<JsonPath> parsed_paths;
-        const char* cstr;
-        auto err = path.get_c_str().get(cstr);
-        if (err) {
-            return Status::InvalidArgument(strings::Substitute("Invalid json path: $0, code $1, error: $2", jsonpath,
-                                                               err, simdjson::error_message(err)));
+            JsonFunctions::parse_json_paths(std::string(cstr), &parsed_paths);
+            path_vecs->push_back(parsed_paths);
```
Comment (sduzh) [inline be/src/exec/vectorized/json_scanner.cpp:226]:
```suggestion
            path_vecs->emplace_back(std::move(parsed_paths));
```

**Review by sduzh — APPROVED** [review]

File: be/src/exec/vectorized/json_scanner.cpp (hunk 3 above)
Comment (rickif) [reply be/src/exec/vectorized/json_scanner.cpp]: Updated.

File: be/src/exec/vectorized/json_scanner.cpp (hunk 4 above)
Comment (rickif) [reply be/src/exec/vectorized/json_scanner.cpp]: Updated.

Comment (rickif) [reply be/src/exec/vectorized/json_scanner.cpp]: Updated.

**Review by chaoyli — APPROVED** [review]

**Pull request merged**